### PR TITLE
BUG: fix TypeErrors raised within _python_agg_general

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @jreback @WillAyd 

There are a few ways in which we incorrectly raise TypeError within _python_agg_general that this fixes.

A lot of the complexity in this code comes from the fact that we drop columns on which a function is invalid instead of requiring the user to subset columns.

### CODE REVIEW

**Comment by WillAyd** [comment]
This does conflict with #29124 - wonder if that would have any material impact on exception catching here (maybe, maybe not)

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -899,10 +899,21 @@ def _python_agg_general(self, func, *args, **kwargs):
         output = {}
         for name, obj in self._iterate_slices():
             try:
-                result, counts = self.grouper.agg_series(obj, f)
+                # if this function is invalid for this dtype, we will ignore it.
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: hmm, this makes us potentially run the function twice?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py]: its the same pattern we use in libreduction to call the function on any empty Series

Comment (jreback) [reply pandas/core/groupby/groupby.py]: k cool, yeah I think we have some tests for this, interesting they didn't break

**Review by jreback — APPROVED** [review]

**Pull request merged**